docs/book/tutorials/enabling-vsphere-csi-on-an-existing-cluster.md

```diff
@@ -0,0 +1,133 @@
+
+Finally, the `disk.EnableUUID` parameter must be set for each node VMs. This step is necessary so that the VMDK always presents a consistent UUID to the VM, thus allowing the disk to be mounted properly.
+It is recommended to not take snapshots of CNS node VMs to avoid errors and unpredictable behavior.
+
+#### disk.EnableUUID=1
+
+The following govc commands will set the disk.EnableUUID=1 on all nodes.
+
+```sh
+export GOVC_INSECURE=1
+export GOVC_URL='https://<VC_IP>'
+export GOVC_USERNAME=VC_Admin_User
+export GOVC_PASSWORD=VC_Admin_Passwd
+```
+
+Check the connection to vCenter:
+
+```sh
+$ govc ls
+/datacenter/vm
+/datacenter/network
+/datacenter/host
+/datacenter/datastore
+```
+
+To retrieve all Node VMs, use the following command:
+
+```sh
+$ govc ls /<datacenter-name>/vm
+/datacenter/vm/k8s-node3
+/datacenter/vm/k8s-node4
+/datacenter/vm/k8s-node1
+/datacenter/vm/k8s-node2
+/datacenter/vm/k8s-master
+```
+
+To use govc to enable Disk UUID, use the following command:
+
+```sh
+govc vm.change -vm '/datacenter/vm/k8s-node1' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node2' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node3' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-node4' -e="disk.enableUUID=1"
+govc vm.change -vm '/datacenter/vm/k8s-master' -e="disk.enableUUID=1"
+```
+
+Further information on disk.enableUUID can be found in [VMware Knowledgebase Article 52815](https://kb.vmware.com/s/article/52815).
+
+#### Upgrade Virtual Machine Hardware
+
+VM Hardware should be at version 15 or higher.
+
+```bash
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node1'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node2'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node3'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-node4'
+govc vm.upgrade -version=15 -vm '/datacenter/vm/k8s-master'
+```
+
+Check the VM Hardware version after running the above command:
+
+```bash
+$ govc vm.option.info '/datacenter/vm/k8s-node1' | grep HwVersion
+HwVersion:           15
+```
+
+## Kubernetes changes
+
+### Node-level changes
+
+On each K8s node, set the `kubelet`’s `cloud-provider` flag to `external` on all nodes. This flag needs to be set in the service configuration file (usually `/etc/systemd/system/kubelet.service`) but this depends on how you installed Kubernetes or the distribution you are using.
+
+E.g:
+
+```sh
+--cloud-provider=external
+```
+
+Restart the `kubelet` service on each node.
+
+```sh
+systemctl daemon-reload
+systemctl restart kubelet.service
+```
+
+### Kubernetes manifest changes
+
+Set taints on all nodes to allow them to be initialised by the vSphere Cloud Provider Interface, this allows them to have their `providerID` populated, which creates the link between the CSI and the VM in vCenter.
+
+On worker nodes set this taint:
+
+```sh
+kubectl taint nodes <your k8s node name> node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
```

Comment:
Maybe not a bad idea to have a command to taint all nodes at once. Assuming CPI is installed correctly it should reconcile things pretty quickly.

Reply:
You thinking like this? 

```sh
kubectl taint nodes --all node.cloudprovider.kubernetes.io/uninitialized=true:NoSchedule
```

Reply:
You definitely need to taint both master and worker nodes. So we need to have the note/instruction for both.